### PR TITLE
Built-in variables

### LINKED ISSUE #55: Lacking built-in variables
We need to be able to provide built-in variables (https://iver-wharf.github.io/#/usage-wharfyml/variables/built-in-variables) in both `.wharf-ci.yml` parsing as a variable substitution source as well as in the step types as some have built-in var depending default values, such as `docker.group` step type field (https://iver-wharf.github.io/#/usage-wharfyml/step-types/docker) that should default to `REPO_GROUP`.

Some variables are more environment focused, such as:

- `REG_URL`
- `DEFAULT_DOMAIN`
- `CHART_REPO`

While others are build-specific, such as:

- `BUILD_REF`
- `GIT_COMMIT`
- `REPO_GROUP`

Need to plan out how to provide these built-in vars in a good way that works both locally and from the Kubernetes perspective.

One idea is to have `.env` files of values, where we could let wharf-cmd look at multiple and merge the values.

Such as:

```env
# ~/.config/iver-wharf/wharf-cmd/built-in-vars.env
CHART_REPO=https://harbor.local/library
```

```env
# /tmp/iver-wharf/wharf-cmd/build/123/built-in-vars.env
BUILD_REF=123
GIT_COMMIT=9a634cb05824565110a119bd60ef6e1c8243bb40
```

Suggest then that wharf-cmd looks at:

- `~/.config/iver-wharf/wharf-cmd/built-in-vars.env`
- `/etc/iver-wharf/wharf-cmd/built-in-vars.env`
- All `.wharf-cmd-built-in-vars.env` files in parent directories, recursively

And uses those to auto-generate `/tmp/iver/wharf-cmd/build/123/built-in-vars.env`, which would be what it's actually using in the build.

With this, we could provide a flag such as `--var-env-file=file.env` that adds another place for Wharf to look for these files, and let the variables be populated by Kubernetes secrets and whatnot.

Kubernetes secrets and configmaps are mounted as each file as a separate secret value, so perhaps add support for that instead of forcing `.env` file would be a good idea.

Also, YAML is probably a better format than `.env` anyway, as it preserves the types.

---

If it wasen't clear from the rambling issue description, this definetly needs some dedicated planning time before coding starts.

### PR DESCRIPTION
**Pull request by jilleJr**
- \[ ] I've added a new note in the `CHANGELOG.md` file, according to docs:
  https://iver-wharf.github.io/#/development/changelogs/writing-changelogs

## Summary

- Added builtin vars file listing inside new `pkg/wharfyml/builtins.go` files
- Added `varsub.Source` all the way down to step type parsing inside `pkg/wharfyml`
- Added current dir-evaluating based on where the `.wharf-ci.yml` file is
- Added defaults for steps based on built-in vars
- Added gitstats as `varsub.Source`

## Motivation

There was a big discussion in #55, and this implementation adds what we concluded.

The same conclusions can also be seen in the logs if you get "missing built-in var" error:

```console
$ go run . run -p test/.wharf-ci.yml
[WARN ] Cannot run build due to parsing errors.  errors=3
[WARN ]    4:5   build/my-build/docker: missing built-in var: need REG_URL or "registry" to construct "destination"
[WARN ]   11:5   helm-pack/my-chart/helm-package: missing built-in var: need CHART_REPO to construct "destination"
[WARN ]   16:5   helm-deploy/my-chart/helm: missing built-in var: need CHART_REPO to construct "repo"
[INFO ] Tip: You can add built-in variables to Wharf in multiple ways.
	
	Wharf look for values in the following files:
	  /etc/iver-wharf/wharf-cmd/wharf-vars.yml
	  ~/.config/iver-wharf/wharf-cmd/wharf-vars.yml
	
	Wharf also looks for:
	  - All ".wharf-vars.yml" in this directory or any parent directory.
	  - Local Git repository and extracts GIT_ and REPO_ variables from it.
	  - Environment variables.
	
	Sample file content:
	  # .wharf-vars.yml
	  vars:
	    REG_URL: http://harbor.example.com
	
[ERROR] failed to parse .wharf-ci.yml
exit status 1
```

This implementation is made so that you could write your .wharf-ci.yml file to bypass all the built-in variables. They are only there to help.

In other words, this makes the `docker.destination` field required, iff `REG_URL`, `REPO_NAME`, and `REPO_GROUP` are undefined. Same deal for other fields of some of the other step types.

Closes #55

### CODE REVIEW

File: pkg/wharfyml/nodemapparser.go [hunk 1]
```diff
@@ -147,3 +176,49 @@ func (p nodeMapParser) newRequiredError(key string) error {
 	inner := fmt.Errorf("%w: %q", ErrMissingRequired, key)
 	return wrapPosErrorNode(inner, p.parent)
 }
+
+func (p nodeMapParser) readFromVarSubForOther(
```
Comment (Alexamakans) [inline pkg/wharfyml/nodemapparser.go:180]: The bulk of this method is a duplicate of `lookupFromVarSubForOther`.

With the slight difference of saying `"%w: need %varLookup% or %nodeKey% to construct %other%"` for the first error instead of `"%w: need %varLookup% to construct %other%"`, line #185 to #197 is the same as `lookupFromVarSubForOther`.

Would be nice if the error could remain as-is but we could call `lookupFromVarSubForOther` in this method.

Maybe something like ( Diff here: https://www.textcompare.org/?id=6233241f197ed215ecce4900 )
```go
func (p nodeMapParser) readFromVarSubForOther(
	nodeKey, varLookup, other string, source varsub.Source) error {
	if _, ok := p.nodes[nodeKey]; ok {
		return nil
	}
	newNode, err := p.lookupFromVarSubForOther(varLookup, other, source, "or", nodeKey)
	p.nodes["registry"] = newNode
	return err
}

func (p nodeMapParser) lookupFromVarSubForOther(
	varLookup, other string, source varsub.Source, additionalErrs ...string) (*yaml.Node, error) {
	value, ok := safeLookupVar(source, varLookup)
	if !ok {
		err := fmt.Errorf("%w: need %s to construct %q",
			ErrMissingBuiltinVar, strings.Join(append([]string{varLookup}, additionalErrs...), " "), other)
		return nil, wrapPosErrorNode(err, p.parent)
	}
	newNode, err := newNodeWithValue(p.parent, value)
	if err != nil {
		err := fmt.Errorf("read %s to construct %q: %w",
			varLookup, other, err)
		return nil, wrapPosErrorNode(err, p.parent)
	}
	return newNode, nil
}
```
?

Not the best, but it reads a lot better IMO.

**Review by Alexamakans — CHANGES_REQUESTED** [review]
Very nice! 👍🏻

Had a couple of confusing moments, and that's mainly what the comments are about.

File: CHANGELOG.md [hunk 2]
```diff
@@ -50,6 +50,25 @@ This project tries to follow [SemVer 2.0.0](https://semver.org/).
...
+- Added support for a new file type: `.wharf-vars.yml`. It is used to define
+  built-in variables, and wharf-cmd looks for it at: (#73)
+
+  - `.wharf-vars.yml` in same directory as `.wharf-ci.yml`
+  - `.wharf-vars.yml` in all parent directories, recursively.
+  - Linux: `/etc/.config/iver-wharf/wharf-cmd/wharf-vars.yml`
+  - Linux: `~/.config/iver-wharf/wharf-cmd/wharf-vars.yml`
+  - Mac: `~/Library/Application Support/iver-wharf/wharf-cmd/wharf-vars.yml`
+  - Windows: `%APPDATA%\iver-wharf\wharf-cmd\wharf-vars.yml`
+
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply CHANGELOG.md:62]: Would prefer to reuse the bullet point list for showing the priority, as long paragraphs are not that glanceable.

Changed to:

> - Added support for a new file type: `.wharf-vars.yml`. It is used to define
>   built-in variables, and wharf-cmd looks for it in multiple files in the
>   following order, where former files take precedence over latter files on a
>   per-variable basis: (#73)
> 
>   - `./.wharf-vars.yml` (in same directory as `.wharf-ci.yml`)
>   - `./../.wharf-vars.yml` (in parent directory of `.wharf-ci.yml`)
>   - `./../../.wharf-vars.yml` (etc; it continues recursively)
>   - (Linux only) `~/.config/iver-wharf/wharf-cmd/wharf-vars.yml`
>   - (Linux only) `/etc/.config/iver-wharf/wharf-cmd/wharf-vars.yml`
>   - (Darwin/OS X only) `~/Library/Application Support/iver-wharf/wharf-cmd/wharf-vars.yml`
>   - (Windows only) `%APPDATA%\iver-wharf\wharf-cmd\wharf-vars.yml`

What do you think?

File: CHANGELOG.md [hunk 3]
```diff
@@ -50,6 +50,25 @@ This project tries to follow [SemVer 2.0.0](https://semver.org/).
...
 
+- Added support for a new file type: `.wharf-vars.yml`. It is used to define
+  built-in variables, and wharf-cmd looks for it at: (#73)
+
+  - `.wharf-vars.yml` in same directory as `.wharf-ci.yml`
+  - `.wharf-vars.yml` in all parent directories, recursively.
+  - Linux: `/etc/.config/iver-wharf/wharf-cmd/wharf-vars.yml`
+  - Linux: `~/.config/iver-wharf/wharf-cmd/wharf-vars.yml`
+  - Mac: `~/Library/Application Support/iver-wharf/wharf-cmd/wharf-vars.yml`
+  - Windows: `%APPDATA%\iver-wharf\wharf-cmd\wharf-vars.yml`
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply CHANGELOG.md:61]: Good idea. I've added the following:

> Note the leading dot in the directory tree files (`.wharf-vars.yml`), while
> the files from config folders is without the dot (`wharf-vars.yml`).

And yea I'd say it's pretty conventional. Or at least it's very unconventional to have dotfiles inside a config directory.

The closest comparison I could think of is the Git config files, which is either stored as `~/.gitconfig` (dotfile) or `.git/config` (no dot).

The argument of "it's already hidden" is the selling point for me. It's just weird to have dotfiles inside a config directory.

File: CHANGELOG.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Alexamakans) [reply CHANGELOG.md:68]: Nice 👍🏻

File: pkg/wharfyml/nodemapparser.go (hunk 1 above)
Comment (jilleJr) [reply pkg/wharfyml/nodemapparser.go]: Fixed!

I blame myself for coding while I was so tired.

File: pkg/wharfyml/nodemapparser.go [hunk 4]
```diff
@@ -147,3 +176,49 @@ func (p nodeMapParser) newRequiredError(key string) error {
...
+			"read %s to construct %q: %w", varLookup, other, err),
+			p.parent)
+	}
+	p.nodes["registry"] = newNode
+	return nil
+}
+
+func (p nodeMapParser) lookupFromVarSubForOther(
+	varLookup, other string, source varsub.Source) (*yaml.Node, error) {
+	repoNameVar, ok := safeLookupVar(source, varLookup)
```
[the remark this answers is not in the document]
Comment (jilleJr) [reply pkg/wharfyml/nodemapparser.go]: Yep, it's a rememant of my first iteration that was hardcoded to repoName, similar to the above misstake of `p.nodes["registry"]`

Nice catch! Fixed

File: pkg/wharfyml/nodemapparser.go [hunk 5]
```diff
@@ -111,20 +111,20 @@ func (p nodeMapParser) unmarshalStringStringMap(key string, target *map[string]s
...
 }
 
-func (p nodeMapParser) unmarshalStringFromVarSubForOther(
-	varLookup, other string, source varsub.Source, target *string) error {
-	node, err := p.lookupFromVarSubForOther(
-		varLookup, other, source)
+func (p nodeMapParser) unmarshalStringFromVarSub(
+	varLookup string, source varsub.Source, target *string) error {
+	node, err := p.lookupFromVarSub(
+		varLookup, source)
```
Comment (Alexamakans) [inline pkg/wharfyml/nodemapparser.go:127]:
```suggestion
	node, err := p.lookupFromVarSub(varLookup, source)
```

**Review by Alexamakans — APPROVED** [review]
Byodiful
